cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -0,0 +1,274 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Device object representing Google devices with a grid qubit layout."""
+
+from typing import Any, Set, Tuple, cast
+import cirq
+from cirq_google.api import v2
+
+
+def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) -> None:
+    """Validates the DeviceSpecification proto.
+
+    Args:
+        proto: The DeviceSpecification proto to validate.
+
+    Raises:
+        ValueError: If the DeviceSpecification is invalid.
+
+    """
+
+    for target_set in proto.valid_targets:
+
+        # Check for unknown qubits in targets.
+        for target in target_set.targets:
+            for target_id in target.ids:
+                if target_id not in proto.valid_qubits:
+                    raise ValueError(
+                        f"Invalid DeviceSpecification: valid_targets contain qubit '{target_id}'"
+                        " which is not in valid_qubits."
+                    )
+
+        # Symmetric and asymmetric targets should not have repeated qubits.
+        if (
+            target_set.target_ordering == v2.device_pb2.TargetSet.SYMMETRIC
+            or target_set.target_ordering == v2.device_pb2.TargetSet.ASYMMETRIC
+        ):
+            for target in target_set.targets:
+                if len(target.ids) > len(set(target.ids)):
+                    raise ValueError(
+                        f"Invalid DeviceSpecification: the target set '{target_set.name}' is either"
+                        " SYMMETRIC or ASYMMETRIC but has a target which contains repeated qubits:"
+                        f" {target.ids}."
+                    )
+
+        # A SUBSET_PERMUTATION target should contain exactly one qubit.
+        if target_set.target_ordering == v2.device_pb2.TargetSet.SUBSET_PERMUTATION:
+            for target in target_set.targets:
+                if len(target.ids) != 1:
+                    raise ValueError(
+                        f"Invalid DeviceSpecification: the target set '{target_set.name}' is of"
+                        " type SUBSET_PERMUTATION but contains a target which does not have exactly"
+                        f" 1 qubit: {target.ids}."
+                    )
+
+
+@cirq.value_equality
+class GridDevice(cirq.Device):
+    """Device object representing Google devices with a grid qubit layout.
+
+    For end users, instances of this class are typically accessed via
+    `Engine.get_processor('processor_name').get_device()`.
+
+    This class is compliant with the core `cirq.Device` abstraction. In particular:
+        * Device information is captured in the `metadata` property.
+        * An instance of `GridDevice` can be used to validate circuits, moments, and operations.
+
+    Example use cases:
+
+        * Get an instance of a Google grid device.
+        >>> device = cirq_google.get_engine().get_processor('weber').get_device()
```

Comment:
I wonder if we should create an example that does not require access to the API.  This could be a follow-up issue/PR if this is not yet done.  For instance, using the virtual engine stuff that we are building.